### PR TITLE
python312Packages.grep-ast: init at 0.3.3

### PR DESCRIPTION
**Pull request by greg-hellings**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by drupol — CHANGES_REQUESTED** [review]
I tried this PR and it's failing on `x86_64-linux`:

```
❯ nix build .#python3Packages.grep-ast -L
error:
       … while evaluating the attribute 'drvPath'
         at /nix/store/shg5a9h51v2vbiy81hljbf1wmr65rh6k-source/lib/customisation.nix:365:7:
          364|     in commonAttrs // {
          365|       drvPath = assert condition; drv.drvPath;
             |       ^
          366|       outPath = assert condition; drv.outPath;

       … while evaluating the attribute 'drvPath'
         at /nix/store/shg5a9h51v2vbiy81hljbf1wmr65rh6k-source/lib/customisation.nix:365:7:
          364|     in commonAttrs // {
          365|       drvPath = assert condition; drv.drvPath;
             |       ^
          366|       outPath = assert condition; drv.outPath;

       … while calling the 'derivationStrict' builtin
         at /builtin/derivation.nix:1:208:
       (stack trace truncated; use '--show-trace' to show the full trace)

       error: tree-sitter0_21-0.21.3 not supported for interpreter python3.12
~/C/N/nixpkgs > init-grep-ast ✘   
```

**Comment by greg-hellings** [comment]
Hmm, so this is a sticky wicket. I developed this against nixos-24.05 initially. Looks like what we're uncovering is a few things:

* This package depends on tree-sitter-languages, which was just added to the current nixos-unstable. However, it depends on tree-sitter0_21. That is pinned to < Python 3.12 due to dependencies on distutils.
* tree-sitter 0.21.1 allowed using it for >= Python 3.12, if distutils is transmitted with the package as a dependency. The version in nixpkgs is 0.21.3 already, but the nix `disabled` flag for the tree-sitter0_21 package still limits it to only building for < 3.12

What I don't know is: is it appropriate for me to make that change to tree-sitter0_21 as part of this PR? It would unblock this to allow it to build on Python 3.12. It is a relatively small change. I'm going to push it up with this branch. Please let me know if it is not appropriate to make that change as part of this PR.

All of this is a problem because the tree-sitter-languages package is unmaintained upstream, despite being newly added to nix in the current cycle.

File: pkgs/development/python-modules/tree-sitter0_21/default.nix
Comment (drupol) [inline pkgs/development/python-modules/tree-sitter0_21/default.nix:37]: How about making it optional for python >= 3.12?

Comment (greg-hellings) [reply pkgs/development/python-modules/tree-sitter0_21/default.nix:37]: distutils is required for tree-sitter 0.21 to work. It was removed from the core library in Python 3.12, so it's not really an optional dependency here.

Comment (greg-hellings) [reply pkgs/development/python-modules/tree-sitter0_21/default.nix:36]: I suppose this could be something like 

```
if pythonAtLeast "3.12" then [ distutils ] else []
```

if we needed to. But including the distutils works just fine in 3.11. It results in more or less a noop, since distutils is part of the core Python packaging up until 3.12 when it was removed.

Comment (drupol) [reply pkgs/development/python-modules/tree-sitter0_21/default.nix:36]:
```nix
propagatedBuildInputs = lib.optionals (pythonAtLeast "3.12") [ distutils ];
```

**Comment by greg-hellings** [comment]
This one appears to be through checks successfully and I've addressed the PR comments.

**Comment by drupol** [comment]
@ofborg build python3Packages.grep-ast

File: pkgs/development/python-modules/grep-ast/default.nix
Comment (GaetanLepage) [inline pkgs/development/python-modules/grep-ast/default.nix:13]:
```suggestion
  version = "0.3.3";
  pyproject = true;

```

File: pkgs/development/python-modules/grep-ast/default.nix
Comment (GaetanLepage) [inline pkgs/development/python-modules/grep-ast/default.nix:20]:
```suggestion
  dependencies = [
```

File: pkgs/development/python-modules/grep-ast/default.nix
Comment (GaetanLepage) [inline pkgs/development/python-modules/grep-ast/default.nix:28]:
```suggestion
    maintainers = with lib.maintainers; [ greg ];
```

File: pkgs/development/python-modules/tree-sitter0_21/default.nix
Comment (GaetanLepage) [inline pkgs/development/python-modules/tree-sitter0_21/default.nix:35]:
```suggestion
  dependencies = lib.optionals (pythonAtLeast "3.12") [ distutils ];
```

**Review by GaetanLepage — DISMISSED** [review]

File: pkgs/development/python-modules/grep-ast/default.nix
Comment (greg-hellings) [reply pkgs/development/python-modules/grep-ast/default.nix:13]: It isn't a pyproject-based library.

Comment (GaetanLepage) [reply pkgs/development/python-modules/grep-ast/default.nix:13]: Yes, but it should work nonetheless. It's counter intuitive I agree.

Comment (greg-hellings) [reply pkgs/development/python-modules/grep-ast/default.nix:13]: It doesn't, though. I get this error.

<details>
```
greg@jude ~/s/nixpkgs init-grep-ast @ nix build ".#python3Packages.grep-ast"
warning: Git tree '/home/greg/src/nixpkgs' is dirty
error: builder for '/nix/store/0ybl3jji0zvn2pkxmyz6mis7c4bdw6xv-python3.12-grep-ast-0.3.3.drv' failed with exit code 1;
       last 10 log lines:
       >   File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
       >   File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
       >   File "<frozen importlib._bootstrap>", line 1310, in _find_and_load_unlocked
       >   File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
       >   File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
       >   File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
       >   File "<frozen importlib._bootstrap>", line 1324, in _find_and_load_unlocked
       > ModuleNotFoundError: No module named 'setuptools'
       >
       > ERROR Backend 'setuptools.build_meta:__legacy__' is not available.
       For full logs, run 'nix log /nix/store/0ybl3jji0zvn2pkxmyz6mis7c4bdw6xv-python3.12-grep-ast-0.3.3.drv'.
```
</details>

Comment (GaetanLepage) [reply pkgs/development/python-modules/grep-ast/default.nix:13]: Yes, add `build-system = [ setuptools ];`

File: pkgs/development/python-modules/grep-ast/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/grep-ast/default.nix:24]: Could you enable tests with `pythonImportsCheck` and `pytestCheckHook`?
https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#testing-python-packages-testing-python-packages

File: pkgs/development/python-modules/grep-ast/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/grep-ast/default.nix:25]: Could you add `meta.mainProgram`?

File: pkgs/development/python-modules/tree-sitter0_21/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/tree-sitter0_21/default.nix:35]: Please don't propagate `distutils`. It should go to `build-system` instead.

**Review by natsukium — CHANGES_REQUESTED** [review]
I think this is an application but not a library. Please consider moving it to `pkgs/by-name/gr/grep-ast`.

File: pkgs/development/python-modules/tree-sitter0_21/default.nix
Comment (greg-hellings) [reply pkgs/development/python-modules/tree-sitter0_21/default.nix:35]: Without distutils being propagated to downstream libraries, tree-sitter 0.21 does not work on Python 3.12. It isn't part of this package's build system, but rather part of the library's internal calls that other libraries using this one depend on. Removing it from dependencies and putting it in build-system would be incorrect and render this package non-functional again.

**Comment by greg-hellings** [comment]
> I think this is an application but not a library. Please consider moving it to `pkgs/by-name/gr/grep-ast`.

It is a library. It comes with an application script entry point, but I'm actually only interested in it as a library upon which the package I am trying to get into nixpkgs depends.

File: pkgs/development/python-modules/grep-ast/default.nix
Comment (greg-hellings) [reply pkgs/development/python-modules/grep-ast/default.nix:29]: Done

File: pkgs/development/python-modules/grep-ast/default.nix
Comment (greg-hellings) [reply pkgs/development/python-modules/grep-ast/default.nix:34]: Done

File: pkgs/development/python-modules/grep-ast/default.nix
Comment (greg-hellings) [reply pkgs/development/python-modules/grep-ast/default.nix]: Done

**Review by drupol — DISMISSED** [review]
Could you please rename the PR to: `python3Packages.grep-ast: init at 0.3.3` and also make sure there's only one commit in it so that the builds are automatically triggered by the CI ?

File: pkgs/development/python-modules/tree-sitter0_21/default.nix
Comment (natsukium) [reply pkgs/development/python-modules/tree-sitter0_21/default.nix:35]: Weird. I wonder why it works with python<=3.11 even though setuptools was not propagated?
Well, this time, it's okay.

**Comment by natsukium** [comment]
> Could you please rename the PR to: python3Packages.grep-ast: init at 0.3.3

This is not necessary. Instead, we recommend `python312Packages`.
See https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#contributing-guidelines-contributing-guidelines

> Commit names of Python libraries should reflect that they are Python libraries, so write for example python311Packages.numpy: 1.11 -> 1.12. It is highly recommended to specify the current default version to enable automatic build by ofborg. Note that pythonPackages is an alias for python27Packages.

**Review by natsukium — APPROVED** [review]
`tree-sitter0_21` must not be propagated as a dependency at all, but it is outside the scope of this PR.
I think we can merge it when the commits are squashed.

File: pkgs/development/python-modules/tree-sitter0_21/default.nix
Comment (greg-hellings) [reply pkgs/development/python-modules/tree-sitter0_21/default.nix:35]: distutils was still included in the base library up through 3.11. It has been deprecated since forever, but it wasn't removed from the stdlib until 3.12. Since this library uses distutils internally (up through version 0.21) with direct calls, it's needed to bring it forward with this version of tree-sitter.

**Comment by greg-hellings** [comment]
Squash merge completed

**Comment by drupol** [comment]
Please read the contributing guide to learn how to write commit log messages.

**Comment by greg-hellings** [comment]
> Please read the contributing guide to learn how to write commit log messages.

I have read them many times. What are you objecting to in the message?

**Comment by drupol** [comment]
> > Please read the contributing guide to learn how to write commit log messages.
> 
> I have read them many times. What are you objecting to in the message?

You did it correctly here: https://github.com/NixOS/nixpkgs/pull/331174/commits why one would want to make it different here?

**Comment by greg-hellings** [comment]
> > > Please read the contributing guide to learn how to write commit log messages.
> > 
> > 
> > I have read them many times. What are you objecting to in the message?
> 
> You did it correctly here: [#331174 (commits)](https://github.com/NixOS/nixpkgs/pull/331174/commits) why one would want to make it different here?

Ah, I see the discrepancy. That should be fixed up.

File: pkgs/development/python-modules/grep-ast/default.nix
Comment (Sigmanificient) [inline pkgs/development/python-modules/grep-ast/default.nix:38]: *`meta.description` must: [...] not start with the definite or an indefinite article.*
> https://github.com/NixOS/nixpkgs/tree/master/pkgs#meta-attributes

```suggestion
    description = "Python interface to the ast-grep tool";
```

**Review by Sigmanificient — CHANGES_REQUESTED** [review]
nit

**Review by eclairevoyant — DISMISSED** [review]
Please split into two commits:

* python312Packages.grep-ast: init at 0.3.3
* python312Packages.tree-sitter0_21: remove Python 3.12 limit

as natsukium asked earlier, and address sigmanificient's feedback

File: pkgs/development/python-modules/grep-ast/default.nix
Comment (greg-hellings) [reply pkgs/development/python-modules/grep-ast/default.nix]: Done

**Comment by greg-hellings** [comment]
> Please split into two commits:
> 
>     * python312Packages.grep-ast: init at 0.3.3
> 
>     * python312Packages.tree-sitter0_21: remove Python 3.12 limit
> 
> 
> as natsukium asked earlier, and address sigmanificient's feedback

Done

**Review by Sigmanificient — APPROVED** [review]

**Comment by eclairevoyant** [comment]
it should be python312Packages not python3Packages (for the reason natsukium mentioned earlier, and per https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#contributing-guidelines-contributing-guidelines)

**Comment by greg-hellings** [comment]
Whoops! Fixed.

**Comment by fabaff** [comment]
Result of `nixpkgs-review pr 331173` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages failed to build:</summary>
  <ul>
    <li>browsr</li>
    <li>browsr.dist</li>
    <li>python312Packages.graphrag</li>
    <li>python312Packages.graphrag.dist</li>
    <li>textual-paint</li>
    <li>textual-paint.dist</li>
  </ul>
</details>
<details>
  <summary>32 packages built:</summary>
  <ul>
    <li>frogmouth</li>
    <li>frogmouth.dist</li>
    <li>oterm</li>
    <li>oterm.dist</li>
    <li>python311Packages.grep-ast</li>
    <li>python311Packages.grep-ast.dist</li>
    <li>python312Packages.grep-ast</li>
    <li>python312Packages.grep-ast.dist</li>
    <li>python312Packages.lsp-tree-sitter</li>
    <li>python312Packages.lsp-tree-sitter.dist</li>
    <li>python312Packages.manifestoo</li>
    <li>python312Packages.manifestoo.dist</li>
    <li>python312Packages.pytest-textual-snapshot</li>
    <li>python312Packages.pytest-textual-snapshot.dist</li>
    <li>python312Packages.textual</li>
    <li>python312Packages.textual-dev</li>
    <li>python312Packages.textual-dev.dist</li>
    <li>python312Packages.textual-universal-directorytree</li>
    <li>python312Packages.textual-universal-directorytree.dist</li>
    <li>python312Packages.textual.dist</li>
    <li>python312Packages.tree-sitter-languages</li>
    <li>python312Packages.tree-sitter-languages.dist</li>
    <li>python312Packages.tree-sitter0_21</li>
    <li>python312Packages.tree-sitter0_21.dist</li>
    <li>rich-cli</li>
    <li>rich-cli.dist</li>
    <li>tftui</li>
    <li>tftui.dist</li>
    <li>upiano</li>
    <li>upiano.dist</li>
    <li>wsrepl</li>
    <li>wsrepl.dist</li>
  </ul>
</details>

**Comment by greg-hellings** [comment]
> Result of `nixpkgs-review pr 331173` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> 6 packages failed to build:
> 32 packages built:

Hmm, so these 3 failing packages have all been masked previously from even attempting a build. They all depend either directly or transitively on `tree-sitter-languages` which depends on `tree-sitter0_21`. Thus, they previously weren't even trying to build.

`browsr` build failure is due to a build failure in python312Packages.pymupdf and is not related to these changes.

`textual-paint` was masked by a dependency on [`python312Packages.textual`](https://github.com/NixOS/nixpkgs/pull/332289) which relied directly on `tree-sitter-languages`. But it hard includes its dependencies to outdated versions with internal Python overrides. These include the version of textual that it pulls in, and a build failure in [its internally pinned version of pillow](https://github.com/NixOS/nixpkgs/blob/db15d08107b2ce2e83d49240e9d5dfa5709ec974/pkgs/applications/graphics/textual-paint/default.nix#L12). I will see about creating a separate update for textual-paint once the textual update is merged in.


`python312Packages.graphrag` depends on a version pin of textual and thus isn't building. Once my textual update is merged, I will attempt to get graphrag updated as well. The version in master is currently out of date.

**Comment by greg-hellings** [comment]
How do I handle these 3 failing packages, as they were already failing before my changes? They simply weren't even trying to build, then, because of being disabled with Python 3.12.

**Comment by fabaff** [comment]
If there is an easy way to make them build then the fix would be very welcome in this PR. But this can be a slippery slope and become quickly a rabbit hole as one failure could lead to another and so on and on. Usually it's much faster to fix existing failures later and do the work step by step. In a best case scenario with a simple update.

**Review by fabaff — APPROVED** [review]
Looks good to me.

**Pull request merged**
